openverse_catalog/dags/common/storage/media.py

```diff
@@ -183,6 +189,9 @@ def _initialize_output_path(
             output_file = str(output_file)
         else:
             datetime_string = datetime.now().strftime("%Y%m%d%H%M%S")
+            # If an ingestion date is provided, append it to the filepath
+            if date is not None:
```

Comment:
When testing this locally I encountered filename collisions when two tsvs were created at the same time, since in the reingestion flows we have multiple ingestion flows running concurrently. At the same time, it's not enough to _just_ have the ingestion `date`, because that date will get reingested on subsequent runs. So we definitely need both _somewhere_ in the path.

Looking at this again, I've realized that the current strategy for building the path won't work for reingestion workflows, because we use the [logical_date](https://github.com/WordPress/openverse-catalog/blob/main/openverse_catalog/dags/providers/factory_utils.py#L149), which **is not the same thing as the ingestion date.** Consequently you'll end up with a very confusing file structure. I think I'll make a change to `date_partition_for_prefix` to use `date` when possible, and fall back to `logical_date` (which should always work for non-dated DAGs).

Once that's fixed, we'll still have the collision problem, though. We could pass an `is_reingestion` flag to `date_partition_for_prefix` that we use to indicate the _full_ date should be used as a prefix (eg, `<year>/<month>/<day>/<provider_media-type_timestamp>`. Do you think the resulting folder structure would be too broad, though? I haven't done the math to fully understand how frequently any given day is reingested, so I'm not sure how empty a lot of those folders would be.

Reply:
If not, then I think it would at least be better to do `{date}_{datetime_string}` in the filename instead of the other way around, so that when they're sorted they automatically get grouped by the ingestion date.

Reply:
That's a really great point. I think we'd still want these ultimately grouped together first by logical date then by ingestion date. What do you think? That way for a given run date we'd go to the folder for that run, then `reingestion/<year>/<month>/<day>` or however the partition works for that DAG. Example:
```
/2022/01/17/file_timestamp.tsv
/2022/01/17/reingestion/2022/01/01/file_timestamp.tsv
/2022/01/17/reingestion/2022/01/02/file_timestamp.tsv
```

Reply:
Hm, I suppose it's more likely we'd want to see all the tsvs for a run of the reingestion workflow (so grouped by logical date) than for a particular day being reingested, so that makes sense.

The only thing about fully partitioning by date like that is we'd have a very sparse directory structure. Unless we executed multiple reingestion DagRuns on the same date, then every single `reingestion` folder would have a single file in its sub directories.

What about:
```
# Normal runs are in the wikimedia-commons directory
/year=2022/month=01/day=17/file_timestamp.tsv # A 'normal' run, not reingestion

# Reingestion runs are stored in the wikimedia-reingestion directory, so we already know we're looking at reingestion files
/year=2022/month=01/day=17/2022-01-01/file_timestamp.tsv
/year=2022/month=01/day=17/2022-01-02/file_timestamp.tsv
```

Each of those final directories will contain only one file, but it prevents collisions and at least isn't quite so sparse. Maybe I'm too concerned about avoiding that, it just seems like it would be very annoying to click through 😆